registry/sql-registry/registry/db_registry.py

```diff
@@ -204,6 +204,28 @@ def get_dependent_entities(self, entity_id: Union[str, UUID]) -> List[Entity]:
             downstream_entities, _ = self._bfs(entity_id, RelationshipType.Produces)
         return [e for e in downstream_entities if str(e.id) != str(entity_id)]
     
+    def delete_empty_entities(self, entities: List[Entity]):
+        """
+        Given entity list, deleting all anchors that have no features and all sources that have no anchors.
+        """
+        if len(entities) == 0:
+            return
+        
+        # clean up empty anchors
+        for e in entities:
+            if e.entity_type == EntityType.Anchor:
+                downstream_entities, _ = self._bfs(e.id, RelationshipType.Contains) 
+                if len(downstream_entities) == 1: # only anchor itself
+                    self.delete_entity(e.id)
+        # clean up empty sources
+        for e in entities:
+            if e.entity_type == EntityType.Source:
+                downstream_entities, _ = self._bfs(e.id, RelationshipType.Produces) 
+                if len(downstream_entities) == 1: # only source itself
```

Comment:
Do we need to check the returned entity type? Just to make sure downstream_entities is Anchor/Source?

Reply:
the _bfs entity list return will always contain the entity itself (`e`), which type is already checked above (`e.entity_type`).

Reply:
Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...

Reply:
> Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...

`Source` produces `Anchor`, they have sequence.  

If merged in one loop, for `Source`, you need to check if all the downstream entities has empty `Anchor` which also involves another loop.

But if all empty anchors are cleaned in the first loop, then we can clean empty sources directly.

Reply:
> > Is that better we use one for loop here? Eg. for e in entities: if type is Anchor: ...; elif type is Source: ...
> 
> `Source` produces `Anchor`, they have sequence.
> 
> If merged in one loop, for `Source`, you need to check if all the downstream entities has empty `Anchor` which also involves another loop.
> 
> But if all empty anchors are cleaned in the first loop, then we can clean empty sources directly.

Good point